### ISSUE TITLE
PGP protected reset e-mail

### ISSUE THREAD

**Issue by vdot0x23** [authelia/authelia#965]
I think the ability to send reset emails encrypted to user's public keys and signed with an administrators private key would be a great feature.
Mainly because it allows security/privacy oriented users to distrust their email provider.

I am unfamiliar with React, but I wouldn't mind implementing a backend for an option like "encrypt if PGP key present for user" and "sign if administrative private key present" (pulled from secrets, not ldap due to the least privilege on ldap goal).

Would such a feature be merged?

**Comment by clems4ever** [comment]
Hello @vdot0x23 , I like the idea, I think it is interesting for paranoid users like me not trusting anything :).

So, if I understand well, the users will have their GPG public keys stored in the auth backend (LDAP or file), and one private key (can be one from the admins or a dedicated keypair for Authelia that admins can trust) will be used for signing and exposed as a secret. Now, when an email must be sent for resetting passwords or registering a device, the user gets to check two options before the email is actually sent. If signing is set, the private key stored in secrets is used to sign the email and if encryption is set, the GPG public key of the user the email is destined to is retrieved from auth backend and used for encryption. Have I missed something?

That's sound reasonable to me. What do you think @james-d-elliott and @nightah ?
I'm just a bit worried about the change of workflow in the frontend for people not using that feature. We would need to make sure we don't add useless steps before sending the email for those people. Currently there is no screen, just a popup saying the email has been sent. We need to make sure it will still be the case after the feature is implemented.

**Comment by vdot0x23** [comment]
@clems4ever you got it, very well written :)

To hopefully address your concerns, I see two ways of doing this:
    **a**. Assume the user's public key was inserted during registration
    **b**. Implement registration of a user's public key similar to 2FA registration

If we don't think encrypting device registration emails is needed since they don't grant full access to an account, option **b** is sufficient.

If we do think encrypting device registration email is needed, since they could potentially lock you out of you own account, then option **a** is necessary since we need the key from day one.

**Comment by clems4ever** [comment]
@vdot0x23 , I think we can already make the signature part. Let say we always send the signature if the secret is provided.

For encryption, if it can wait a bit, I think it might be worth waiting for https://github.com/authelia/authelia/issues/303 in order to avoid double work. This feature would allow users to customize their account: add/remove devices, add/remove GPG keys, enable/disable email encryption. What do you think?

**Comment by james-d-elliott** [comment]
I agree with this idea. As long as it's optional and documented. We can implement this into the management UI (user + admin) when it's implemented.

**Comment by vdot0x23** [comment]
@clems4ever I think it will be easier to discuss once I have read the source to Authelia more thoroughly.
If I understand correctly #303 is front-end which I won't touch (unless I accidentally learn React which seems unlikely ;P )

@james-d-elliott Thank you for offering assistance. I just now started playing around with the LDAP suite and I will start tinkering with code soon so it will be a little while :)

**Comment by clems4ever** [comment]
#303 will very likely be a mix of changes in frontend and backend.

**Comment by nightah** [comment]
@vdot0x23 I know you were waiting on a fix to the upstream library to implement this, is this something that you still are interested in contributing?

**Comment by vdot0x23** [comment]
> @vdot0x23 I know you were waiting on a fix to the upstream library to implement this, is this something that you still are interested in contributing?

I decided on skipping the library and doing it by hand (using strings and golang's openpgp), but I'm still having trouble getting email clients to recognize the signature format even though others seem to have suceeded (https://stackoverflow.com/questions/7891263/sign-a-multipart-mail-using-pgp-mime).

I might have made a mistake but didn't expect it to take this long, so I decided to take a break before debuggin it.

**Comment by vdot0x23** [comment]
Okay, I've got a prototype working on (please the messyness, will fix later) https://github.com/vdot0x23/authelia/blob/sign-emails-2/internal/notification/smtp_notifier.go#L211

Decided to drop the library, since it was a little too abstract for this use after all.

Before cleaning up the code, I want to test this (have just been using `thunderbird --file` so far), any suggestions as to how I might do that given the "graphical" nature of verifying both plaintext, html, and signature is present?

**Comment by vdot0x23** [comment]
It seems I can only get this working with either plaintext or HTML, not both :/

Anyway, it may pay to think about how to test this feature (I've been using `thunderbird --file` recently, but that it rather "visual") as this could allow for just trying all kinds of formats (though I have been through quite a few already).

**Comment by james-d-elliott** [comment]
I might be able to take a look at this soon. What did you mean by both? The mixed type emails?

**Comment by vdot0x23** [comment]
> I might be able to take a look at this soon. What did you mean by both? The mixed type emails?

Multipart yes, like this answer: https://stackoverflow.com/questions/7891263/sign-a-multipart-mail-using-pgp-mime

**Comment by james-d-elliott** [comment]
@vdot0x23 sorry to not get back to you sooner. I managed to get this working in a very test driven branch. As you probably experienced it's very touchy about newlines etc. Which was probably the source of your issue.

authelia/authelia:feat-notifier-pgp

https://buildkite.com/authelia/authelia/builds/6441


![image](https://user-images.githubusercontent.com/3903683/126019486-65b5c7de-563f-44fb-801a-282f3a36d0f8.png)

**Comment by james-d-elliott** [comment]
It seems in the time we spent getting this working go has deprecated PGP. This was due in part to the lack of support etc. Do we want to consider moving to the proton mail lib which is effectively a fork of it.. or do we want to stay with the go stdlib version? Or do we want to scrap this idea all together?

**Comment by vdot0x23** [comment]
> It seems in the time we spent getting this working go has deprecated PGP. This was due in part to the lack of support etc. Do we want to consider moving to the proton mail lib which is effectively a fork of it.. or do we want to stay with the go stdlib version? Or do we want to scrap this idea all together?

I still think the idea is very interesting, but as you see from my lack of reply I am rather low on time these days.
The checklist seems reasonable, but I will not be available to implement it in the near future.

**Comment by nightah** [comment]
I'm converting this issue into a discussion for voting purposes.
This feature isn't trivial to implement, and support for it is very scarce.
